test/Microsoft.ML.Tests/Scenarios/Api/AspirationalExamples.cs

```diff
@@ -0,0 +1,61 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.ML.Tests.Scenarios.Api
+{
+    public class AspirationalExamples
+    {
+        public class IrisPrediction
+        {
+            public string PredictedLabel;
+        }
+
+        public class IrisExample
+        {
+            public float SepalWidth { get; set; }
+            public float SepalLength { get; set; }
+            public float PetalWidth { get; set; }
+            public float PetalLength { get; set; }
+        }
+
+        public void FirstExperienceWithML()
+        {
+            // This is the 'getting started with ML' example, how we see it in our new API.
+            // It currently doesn't compile, let alone work, but we still can discuss and improve the syntax.
+
+            // Load the data into the system.
+            string dataPath = "iris-data.txt";
+            var data = TextReader.FitAndRead(env, dataPath, row => (
+                Label: row.ReadString(0),
+                SepalWidth: row.ReadFloat(1),
+                SepalLength: row.ReadFloat(2),
+                PetalWidth: row.ReadFloat(3),
+                PetalLength: row.ReadFloat(4)));
+
+
+            var preprocess = data.Schema.MakeEstimator(row => (
```

Comment:
Why do we need to expose Schema to users here? Can we simply do data.MakeExtimator(row => ...)

Reply:
this is how I originally phrased it, but it makes an appearance that the data itself is being 'used' and somehow 'remembered' by the estimator. Hence the 'Schema' in the middle.

---
In reply to: [208640562](https://github.com/dotnet/machinelearning/pull/639#discussion_r208640562) [](ancestors = 208640562)

Reply:
I see. Now that I think about that, the flow looks a little bit odd in my opinion. You start from the data, then you create an estimator from the data (without using the data) and then you call fit over the data again. In principle you only really need the data at the beginning when you create the reader, right? Or you can actually call fit over different datasets? If the later is true, why do you need to create a reader over a specific dataset at the beginnig? In the current API data and pipeline creations are entangled in multiple places and I think it create confusion a bit.

Reply:
you can actually call `Fit` over different datasets, of course. 

The only thing we really need to `MakeEstimator' is to have the input schema. We sort-of piggy-backed on the `TextReader` here to define the schema and simultaneously define how to read the data file.
We could instead do
```c#
var preprocess = PigstyPipeline.WithSchema(row=> 
row => (
                Label: row.String(),
                SepalWidth: row.Float(),
                SepalLength: row.Float(),
                PetalWidth: row.Float(),
                PetalLength: row.Float())
    .MakeEstimator(row => (Label: row.Label.DictionarizeLabel() ...)
```
in this case, there is no need to read `data` until whenever you want.
But at SOME point you will have to define HOW to read these columns, so lines 30-34 are still needed

---
In reply to: [208660933](https://github.com/dotnet/machinelearning/pull/639#discussion_r208660933) [](ancestors = 208660933)

Reply:
Agreed: at some point you will need to create a reader and define the schema, but I think that separating the 2 (data creation from pipeline creation) is definitely cleaner. Additionally, for the datasets you could only define the schema on the pipeline, and do `env.TextReader.Read(dataPath)` and the validation of the schema will be done on the pipeline at fit time. Basically you could reverse the pattern you have now: define the schema on the pipeline and have the dataset schema matched at fit time, instead of define the schema on the dataset and have the `data.Schema` thing to pass the schema to the pipeline.

Reply:
I don't think this is going to work. You cannot just take an `ISchema` and make text reader configuration of out it: you need to tell what column indices go where. Note the `ReadFloat(0)` and other arguments to `ReadXXX`

---
In reply to: [208726028](https://github.com/dotnet/machinelearning/pull/639#discussion_r208726028) [](ancestors = 208726028)

Reply:
Well but that is a problem of the dataset, not of the pipeline nor of the schema. In theory one would do something like `data.map(row => row._2, row._1, row._0)` if there is the need to change column orders for example. Even because if you have many possible input dataset, some with the right order and some not, it is verbose to repeat the schema definition for all of them, where verbosity == more chances or errors.